### DISCUSSION TITLE
Can we add widget at end of line?

### DISCUSSION THREAD

**Discussion by XiangxiMeng** [microsoft/monaco-editor#3744] [Q&A]
Hi, I was wondering whether we have some api to add widget at end of line, like attached, or you can refer to this link [PlotTool Pro](https://developer.gs.com/discover/products/plottool/)

![gs](https://user-images.githubusercontent.com/60503205/115211459-29ffd180-a132-11eb-9c62-c9915723b349.PNG)

**Comment by CraigMeng** [comment]
we have same requirements here. Have tried overlay widgets, but we cannot figure out how to make widgets scroll with lines

**Comment by acormier** [comment]
Same problem here. 
An overlay widget can be aligned with the right edge of the editor but doesn't scroll with the lines.
A content widget does scroll with the lines but starts at the end of the last character of the line. Can't figure out how to align it to the right edge of the editor.

So, how can I make a widget at the end of a line (associated with that line) but aligned all the way to the editor's right edge and that scrolls with the lines?

Thanks

**Comment by ilias-t** [comment]
try `createDecorationsCollection` instead. something like this worked for me

```ts
 const updateDecorations = () => {
      const model = nextEditor.getModel();
      // clear existing decorations
      nextEditor.createDecorationsCollection([]);
      if (model) {
        const oldDecorations = nextEditor.getDecorationsInRange(model.getFullModelRange()) || [];
        nextEditor.removeDecorations(oldDecorations.map((d) => d.id));

        const decorations = model.getLinesContent().map((_, i) => {
          const lineNumber = i + 1;
          return {
            range: {
              startLineNumber: lineNumber,
              startColumn: 1,
              endLineNumber: lineNumber,
              endColumn: model.getLineMaxColumn(lineNumber),
            },
            options: {
              isWholeLine: true,
              after: {
                content: ` │ ${Math.floor(Math.random() * 10)}h ago • John Doe`,
                inlineClassName: "line-metadata",
                cursorStops: 3,
              },
              stickiness: 1,
              suppressMouseDown: true,
              zIndex: 0,
            },
          } as editor.IModelDeltaDecoration;
        });

        nextEditor.createDecorationsCollection(decorations);
      }
    };

    // Initial decorations
    updateDecorations();

    // Update decorations on content change
    nextEditor.onDidChangeModelContent(updateDecorations);
```


```css
.line-metadata {
  color: var(--gray-9) !important;
  opacity: 0.7;
  font-style: italic;
  pointer-events: none !important;
  user-select: none !important;
}
```

**Reply by ilias-t** [reply]
I haven't tried, but curious if it supports interactive content (like buttons) in that `content` or if it's more like an `::after` pseudo element.
